scene/main/window.cpp

```diff
@@ -1012,7 +1015,9 @@ void Window::popup_centered_ratio(float p_ratio) {
 
 void Window::popup(const Rect2i &p_screen_rect) {
 	emit_signal("about_to_popup");
-
+	if (!is_embedded()) {
+		current_screen = DisplayServer::get_singleton()->window_get_current_screen(get_parent_visible_window()->get_window_id());
+	}
```

Comment:
`current_screen` is already set in this method a little further down. Is there any reason to set it here again?

Reply:
Sorry, but where? I may be blind, but I don't see a current_screen in this method

Reply:
oh I see. Well I fix the logic below then

Reply:
https://github.com/godotengine/godot/blob/a524c10f43ab4e1da0edebec15563ec5b4d0ec2a/scene/main/window.cpp#L1036

Reply:
fixed

Reply:
I should be a bit more awake, when doing changes...
I quickly test with the changes undone

Reply:
now fixed

Reply:
Now `current_screen` gets set right before the popup position is changed, which might result in `current_window` being set, only for the window to be immediately moved to another screen.

Looking at the `DisplayServer` implementations, it seems that whenever a window's bounds are changed, the window gets notified of that through `Window::_rect_changed_callback` which should arguably be the place where the window's `current_screen` should actually be updated. This would make the value of `current_screen` more consistent and also allow `get_current_screen` to use the cached value instead of querying the `DisplayServer` every time.

Reply:
I tested the following cases:
 * current_screen set as in master (in popup) -> windows, like add node, will always beeing placed on main monitor
 * current_screen queried from current_screen in popup -> windows and popups open on the monitor of the active parent window
 * not setting current_screen in popup (only on resize) -> all windows and popups open on first screen. _rect_changed_callback does not provides desired effect, since the current screen is set and the position is updated based on the current_screen. Therefore setting it once in popup with the screen of the active parent window appears to be the only correct fix.
I want to emphasize this last point. You want your windows in your application on the same screen as the window you came from. The application should not just position it where it feels it is useful (i.e. the main screen)

Reply:
Thinking about it a little more, changing `current_screen` only in `_rect_changed_callback` would also create the problem of it only being updated after the `DisplayServer` processes its incoming events, which might be considered a weird behavior to the user who might expect `current_screen` to be updated as soon they call `set_position()`, so perhaps that shouldn't be the only place `current_screen` is set.

That said, are you sure that this isn't fixed by #40922 (which got merged today) or #37506 (which isn't merged yet)? As far as I can tell, `Viewport::_gui_show_tooltip` nor `Window::popup` and the methods it calls don't rely on `current_screen` at any point, unless I'm missing something. How does setting `current_screen` fix this issue?

Reply:
It is not fixed by #40922
It is indeed fixed by #37506, but I will test again once it is rebased. (I don't have the time to rebase it just for fun :))